Cleanup trimming options and document TrimMode=full,partial (#30279)
dotnet/docs: docs/core/deploying/trimming/trimming-options.md

Place this setting in the project file to ensure that the setting applies during `dotnet build`, not just `dotnet publish`.

:::zone pivot="dotnet-7-0"

This setting enables trimming and will trim all assemblies by default. In .NET 6, only assemblies which opted-in

> to trimming via `[AssemblyMetadata("IsTrimmable", "True")]` would be trimmed by default. You can return to the

Author: Good suggestion, but I'm actually not quite sure what we want to do here.

Do we want to define trimmable as anything marked `IsTrimmable=true`? Or do we want to say that it's any assembly that's trimmed?

If it's the former, then `TrimMode=full` trims non-trimmable assemblies. If it's the latter, then there's no direct mapping between `IsTrimmable` and "trimmable".

> to trimming via `[AssemblyMetadata("IsTrimmable", "True")]` would be trimmed by default. You can return to the

Reviewer: I'd say something like "`TrimMode=full` trims all assemblies unless they're explicitly marked as trim-rooted assemblies" and define "trim-rooted" (or whatever term we want).

> to trimming via `[AssemblyMetadata("IsTrimmable", "True")]` would be trimmed by default. You can return to the

Reviewer 2: The suggestion is to define a term that we can use to describe assemblies that are trimmed in TrimMode=partial (or assemblies to which TrimMode applies before .NET 7).

Basically start with saying we have two classes of assemblies and there are controls for what happens to each of the class. And then we have an exception mechanism to treat assemblies from one class as another class.

Once we have the term, we can say:

* < .NET 7 assemblies that are manifested as trimmable get trimmed
* => .NET 7 all assemblies are trimmed, TrimMode=partial is how to get the old behavior

I think it can simplify other parts of the doc too.

> to trimming via `[AssemblyMetadata("IsTrimmable", "True")]` would be trimmed by default. You can return to the

Author: By this definition, in .NET 7 we're trimming "non-trimmable" assemblies. I'm not sure "trimmable" is a good term to use here because I don't think it makes sense without a lot of context.

> to trimming via `[AssemblyMetadata("IsTrimmable", "True")]` would be trimmed by default. You can return to the

Reviewer 2: We cemented ourselves into that term by calling it `IsTrimmable` everywhere (item metadata, the attribute). In .NET 6 we would trim assemblies that are not "trimmable assemblies" if we set TrimmerDefaultAction=link. We could call it differently (trimmable-manifested assemblies?). Or just embrace it...

> to trimming via `[AssemblyMetadata("IsTrimmable", "True")]` would be trimmed by default. You can return to the

Reviewer 2: Or try going with "always-eligible-for-trimming"? (Even that one is not precise if someone goes and sets TrimMode=copy in .NET 6.)

I would probably really just go with "trimmable assembly" and make it obvious that it's not the assemblies that will be trimmed, it's just a name of a category that can be handled differently.

previous behavior by using `<TrimMode>partial</TrimMode>`.

:::zone-end

:::zone pivot="dotnet-6-0,dotnet-5-0"

This setting trims any assemblies that have been configured for trimming. With `Microsoft.NET.Sdk` in .NET 6, this includes any assemblies with `[AssemblyMetadata("IsTrimmable", "True")]`, which is the case for framework assemblies. In .NET 5, framework assemblies from the netcoreapp runtime pack are configured for trimming via `<IsTrimmable>` MSBuild metadata. Other SDKs may define different defaults.

> Starting in .NET 6, this setting also enables the trim-compatibility [Roslyn analyzer](#roslyn-analyzer) and disables [features that are incompatible with trimming](#framework-features-disabled-when-trimming).

Reviewer: > With `Microsoft.NET.Sdk` in .NET 6, this includes any assemblies with `[AssemblyMetadata("IsTrimmable", "True")]`, which is the case for framework assemblies.

Can we be more specific here, as not all frameworks have marked their assemblies as trimmable, e.g. Microsoft.AspNetCore.App still has many untrimmable assemblies.

> Starting in .NET 6, this setting also enables the trim-compatibility [Roslyn analyzer](#roslyn-analyzer) and disables [features that are incompatible with trimming](#framework-features-disabled-when-trimming).

Author: I think this is specific -- the assemblies in `Microsoft.NET.SDK` are mostly marked trimmable. `Microsoft.NET.SDK` does not include `Microsoft.AspNetCore.App`.

> Starting in .NET 6, this setting also enables the trim-compatibility [Roslyn analyzer](#roslyn-analyzer) and disables [features that are incompatible with trimming](#framework-features-disabled-when-trimming).

Reviewer: My contention is with using the word "framework" to implicitly mean `Microsoft.NetCore.App`. My point is that there are other frameworks and we should be clearer.

> This setting trims any assemblies that have been configured for trimming. With `Microsoft.NET.Sdk` in .NET 6, this includes any assemblies with `[AssemblyMetadata("IsTrimmable", "True")]`, which is the case for the .NET core framework assemblies. In .NET 5, framework assemblies from the netcoreapp runtime pack are configured for trimming via `<IsTrimmable>` MSBuild metadata. Other SDKs may define different defaults.

Reviewer 3: ```suggestion
This setting trims any assemblies that have been configured for trimming. With `Microsoft.NET.Sdk` in .NET 6, this includes any assemblies with `[AssemblyMetadata("IsTrimmable", "True")]`, which is the case for the .NET runtime assemblies. In .NET 5, assemblies from the netcoreapp runtime pack are configured for trimming via `<IsTrimmable>` MSBuild metadata. Other SDKs may define different defaults.
```

:::zone-end

## Trimming granularity

:::zone pivot="dotnet-7-0"

The default is to trim all assemblies in the app. This can be changed using the `TrimMode` property.

To only trim assemblies which have opted-in to trimming, set

> ```C#

Reviewer 4: ```suggestion
```csharp
```

Place this setting in the project file to ensure that the setting applies during `dotnet build`, not just `dotnet publish`.

:::zone-end

:::zone pivot="dotnet-7-0"

> This setting enables trimming and will trim all assemblies by default. In .NET 6, only assemblies which opted-in

Reviewer 5: ```suggestion
This setting enables trimming and will trim all assemblies by default. In .NET 6, only assemblies that opted-in
```

to trimming via `[AssemblyMetadata("IsTrimmable", "True")]` would be trimmed by default. You can return to the
previous behavior by using `<TrimMode>partial</TrimMode>`.

:::zone-end

:::zone pivot="dotnet-6-0,dotnet-5-0"

This setting trims any assemblies that have been configured for trimming. With `Microsoft.NET.Sdk` in .NET 6, this includes any assemblies with `[AssemblyMetadata("IsTrimmable", "True")]`, which is the case for the .NET runtime assemblies. In .NET 5, assemblies from the netcoreapp runtime pack are configured for trimming via `<IsTrimmable>` MSBuild metadata. Other SDKs may define different defaults.

:::zone-end

This setting also enables the trim-compatibility [Roslyn analyzer](#roslyn-analyzer) and disables [features that are incompatible with trimming](#framework-features-disabled-when-trimming).

> ## Trimming granularity

Author: Eh, I think the name is fine. "Trimming granularity is an accurate definition of the difference between partial and full.

:::zone pivot="dotnet-7-0"

The default is to trim all assemblies in the app. This can be changed using the `TrimMode` property.

> To only trim assemblies which have opted-in to trimming, set

Reviewer 5: ```suggestion
To only trim assemblies that have opted-in to trimming, set the property to `partial`:
```

```csharp
<TrimMode>partial</TrimMode>
```

> The default setting is

Reviewer 5: ```suggestion
The default setting is `full`:
```

This is equivalent to setting MSBuild metadata `<IsTrimmable>true</IsTrimmable>` for the assembly in `ManagedAssemblyToLink` (see below).

> :::zone-end

Author: Another question is whether we want to add a `NotTrimmableAssembly` ItemGroup.

> :::zone-end

Reviewer: How does one use `TrimmerRootAssembly`? Is it literally like this:

```xml
<ItemGroup>
  <TrimmerRootAssembly Include="AnAssemblyToKeep" />
</ItemGroup>
```

This is equivalent to setting MSBuild metadata `<IsTrimmable>true</IsTrimmable>` for the assembly in `ManagedAssemblyToLink` (see below).

> 
> ## Trimming settings for individual assemblies

Reviewer 4: ```suggestion
## Trimming settings for individual assemblies
```

> This is equivalent to setting MSBuild metadata `<IsTrimmable>true</IsTrimmable>` for the assembly in `ManagedAssemblyToLink` (see below).

Reviewer 5: Does the "see below" still apply? That info doesn't seem to be in the .NET 7 pivot.

Would it be better to just have a table that showed the different values for different versions?

Author: I'll just delete it. We're trying to de-emphasize ManagedAssemblyToLink.

  Control whether the given assembly is trimmed.

:::zone pivot="dotnet-6-0,dotnet-5-0"

- `<TrimMode>copyused</TrimMode>` or `<TrimMode>link</TrimMode>`

  Control the [trimming granularity](#trimming-granularity) of this assembly. This takes precedence over the global `TrimMode`. Setting `TrimMode` on an assembly implies `<IsTrimmable>true</IsTrimmable>`.

:::zone-end

> 

Reviewer 2: Can't comment on the line +7 lines below this one but "All assemblies that do not have `<IsTrimmable>true</IsTrimmable>` are considered roots for the analysis" needs changing too.

  Enable a Roslyn analyzer for a subset of trim analysis warnings.

> ## Warning versions

Author: I'm not sure fully support it. Did we mark every new warning that we added in 7.0 with a new warning version? And add tests for that?

> ## Warning versions

Author: The docs say

>Trim analysis respects the [`AnalysisLevel`](../../project-sdk/msbuild-props.md#analysislevel) property that controls the version of analysis warnings across the SDK

Which isn't true. The intent of warning level is that new warnings (from new SDK versions) don't get added to the old warning version. Basically, warning versions are frozen.

We're not meeting that bar, so we shouldn't advertise it. Still happy to have the option there, but until we properly support it I don't want to put it in the docs.

  Show all detailed warnings, instead of collapsing them to a single warning per assembly.

> The defaults show detailed warnings for the project assembly and `ProjectReference` items. `<TrimmerSingleWarn>` can also be set as metadata on an [individual assembly](#trimmed-assemblies) to control the warning behavior for that assembly only.

Author: Couldn't make it work with the zones. If info about ManagedAssemblyToLink is hidden behind the net6 pivot, the mention can't appear in net7. Also, this doesn't remove all documentation of the option, it just doesn't mention the assembly-specific version when talking about the global property.
